### DISCUSSION TITLE
Measure bond only overlap

### DISCUSSION THREAD

**Discussion by haroldgrosjean** [rdkit/rdkit#4619] [Q&A]
Hello,

I am trying to write a piece of code that automatically detects overlapping bonds between 2 distinct molecules. For example in the image bellow we see that 2 bonds overlap pretty well.

![Screenshot from 2021-10-19 13-49-49](https://user-images.githubusercontent.com/51821930/137914534-3d73bb61-4a28-421b-bbfc-a309c790b993.png)

The files are:
[F199.pdb.txt](https://github.com/rdkit/rdkit/files/7373456/F199.pdb.txt)
[F535.pdb.txt](https://github.com/rdkit/rdkit/files/7373457/F535.pdb.txt)

What I've tried so fare was to build a new empty molecule and add the 2 atoms and the bond from the native molecule and connect them by extracting and re-adding the bond type. I repeat this operation a second time for the other molecules. Then I try to compute the Shape Tanimoto Distance but the molecules I built are obviously achemical and cannot, therefore, be sanitized not can I extract a conformer.

here is the code I have at the moment:
```
import rdkit as rdk
from rdkit import Chem
from rdkit.Chem import AllChem, rdShapeHelpers

F199 = AllChem.MolFromPDBFile('F199.pdb.txt')
F535 = AllChem.MolFromPDBFile('F535.pdb.txt')

#Bonds IDs correspond to one of the overlapping bonds
F199_bond = F199.GetBonds()[2]
F199_A1 = F199_bond.GetBeginAtom()
F199_A2 = F199_bond.GetEndAtom()
mol_bond_F199 = Chem.Mol()
em_F199 = Chem.EditableMol(mol_bond_F199)
em_F199_idx1 = em_F199.AddAtom(F199_A1)
em_F199_idx2 = em_F199.AddAtom(F199_A2)
F199_bond_type = F199_bond.GetBondType()
bond_F199_Idx  = em_F199.AddBond(em_F199_idx1, em_F199_idx2, F199_bond_type)
mol_bond_F199 = em_F199.GetMol()
#also tried mol_bond_F199 = em_F199.GetMol().GetConformers()[0]

F535_bond = F535.GetBonds()[10]
F535_A1 = F535_bond.GetBeginAtom()
F535_A2 = F535_bond.GetEndAtom()
mol_bond_F535 = Chem.Mol()
em_F535 = Chem.EditableMol(mol_bond_F535)
em_F535_idx1 = em_F535.AddAtom(F535_A1)
em_F535_idx2 = em_F535.AddAtom(F535_A2)
F535_bond_type = F535_bond.GetBondType()
bond_F535_Idx  = em_F535.AddBond(em_F535_idx1, em_F535_idx2, F535_bond_type)
mol_bond_F535 = em_F535.GetMol()
#also tried mol_bond_F535 = em_F535.GetMol().GetConformers()[0]

#Breaks here and unsure how to continue
STD = rdShapeHelpers.ShapeTanimotoDist(mol_bond_F199, mol_bond_F535)
```

I would be hugely useful if anyone could indicate me how to carry on this operation as I never explored this part of the rdkit functionalities.

Many thanks in advance,

Bests,

Harold

**Comment by dehaenw** [comment]
Hello @haroldgrosjean ,
one idea you could try is to combine both mols into one mol, then calculate the 3d distance matrix all the atoms, and then see which ones are below a distance threshold, and which ones of those have a bond between them. below my very rough test implementation. i also added a bit to merge the molecules, as i suspect this might be why you are looknig for overlapping bonds. 
```python
import rdkit
from rdkit import Chem
from rdkit.Chem import AllChem, RWMol

F199 = AllChem.MolFromPDBFile('F199.pdb.txt')
F535 = AllChem.MolFromPDBFile('F535.pdb.txt')
merged = Chem.CombineMols(F199,F535)
distmatrix = Chem.rdmolops.Get3DDistanceMatrix(merged)
atoms_to_merge = []
for i,row in enumerate(distmatrix):
    for j,dist in enumerate(row):
        if dist>0 and dist<1 and i>j: #get atoms that are within 1 unit of eachother
            atoms_to_merge.append((i,j))
bonds=[]
for i,id1 in enumerate(atoms_to_merge):
    for id2 in atoms_to_merge[i+1:]:
        b0 = merged.GetBondBetweenAtoms(id1[0],id2[0])
        b1 = merged.GetBondBetweenAtoms(id1[1],id2[1])
        if b0!=None:
            bonds.append([b0,b1])
for bond in bonds:
    print("there's an overlap between the bond with idx {} and with idx {}".format(bond[0].GetIdx(),bond[1].GetIdx()))
            
#if you want to merge the molecules             
merged_final = Chem.RWMol(merged)
idxtoremove = []
for pair in atoms_to_merge:
    atom = merged.GetAtomWithIdx(pair[0])
    idxtoremove.append(atom.GetIdx())
    for neighbor in atom.GetNeighbors():
        merged_final.RemoveBond(neighbor.GetIdx(),pair[0])
        merged_final.AddBond(neighbor.GetIdx(),pair[1],merged.GetBondBetweenAtoms(atom.GetIdx(),neighbor.GetIdx()).GetBondType())
idxtoremove.sort(reverse=True) #sort frmo high to low to have no problems when removing atoms
for idx in idxtoremove:
    merged_final.RemoveAtom(idx)
print("merged molecules here:")
display(merged_final)
display(Chem.MolFromSmiles(Chem.MolToSmiles(merged_final)))
```
this outputs the following. flattened molecule also included in last line for clarity
```
there's an overlap between the bond with idx 25 and with idx 4
there's an overlap between the bond with idx 26 and with idx 2
merged molecules here:
```
![image](https://user-images.githubusercontent.com/66372095/138000852-66087ba9-6356-4174-b61b-45751fd4bafa.png)

**Comment by haroldgrosjean** [comment]
Hi @dehaenw!

Thanks that is great - I'll definitely use some of that code in the future! The case I presented above is a very "simple" example and I anticipate that I will have to be able to robustly identify overlapping bonds in the future. One downstream application of such process would be to enumerate merging strategies. For example there are 2 merging solutions here: C-N-C (based on F535. which you did above) but also N-N-C (based on F199). Would you have an idea of how to calculate the overlap between these bonds? Angles would be also useful but I think I got that covered already.

Cheers!

Harold

**Reply by dehaenw** [reply]
I think just the total distance between each overlapping atom in the overlapping bond could be a good proxy for bond overlap. see below for an example:
```python
import rdkit
from rdkit import Chem
from rdkit.Chem import AllChem, RWMol

F199 = AllChem.MolFromPDBFile('F199.pdb.txt')
F535 = AllChem.MolFromPDBFile('F535.pdb.txt')
merged = Chem.CombineMols(F199,F535)
distmatrix = Chem.rdmolops.Get3DDistanceMatrix(merged)
atoms_to_merge = []
distlist= []
for i,row in enumerate(distmatrix):
    for j,dist in enumerate(row):
        if dist>0 and dist<1 and i>j: #get atoms that are within 1 unit of eachother
            atoms_to_merge.append((i,j))
            distlist.append(dist)
bonds=[]
bonddists=[]
for i,id1 in enumerate(atoms_to_merge):
    for j,id2 in enumerate(atoms_to_merge[i+1:]):
        b0 = merged.GetBondBetweenAtoms(id1[0],id2[0])
        b1 = merged.GetBondBetweenAtoms(id1[1],id2[1])
        if b0!=None:
            bonds.append([b0,b1])
            bonddists.append(distlist[i]+distlist[j])
            
for i,bond in enumerate(bonds):
    print("there's an overlap between the bond with idx {} and with idx {}. The summed distance of the atoms is {}".format(bond[0].GetIdx(),bond[1].GetIdx(),round(bonddists[i],3)))
```
which would output
```
there's an overlap between the bond with idx 25 and with idx 4. The summed distance of the atoms is 1.016
there's an overlap between the bond with idx 26 and with idx 2. The summed distance of the atoms is 1.09
```
which you could then also sort according to your needs.
Concerning the different merging strategies: it might be hard to find some general strategy for this giving well behaved mols (and not have broken up aromatic rings, valence problems etc). For example in your case if you switch the pairs as follows, there are some issues with merging aromatic and non aromatic bonds, unless the following is added when the pdbs are converted to mol
```python
Chem.Kekulize(F199,clearAromaticFlags=True)
Chem.Kekulize(F535,clearAromaticFlags=True)
```
For a general solution you can use on a larger amount of fragments probably something cleaner is needed.
So switching the pair indices gives the other molecule like this:
```python
merged_final = Chem.RWMol(merged)
idxtoremove = []
for pair in atoms_to_merge:
    atom = merged.GetAtomWithIdx(pair[1])
    idxtoremove.append(atom.GetIdx())
    for neighbor in atom.GetNeighbors():
        merged_final.RemoveBond(neighbor.GetIdx(),pair[1])
        merged_final.AddBond(neighbor.GetIdx(),pair[0],merged.GetBondBetweenAtoms(atom.GetIdx(),neighbor.GetIdx()).GetBondType())
idxtoremove.sort(reverse=True) #sort frmo high to low to have no problems when removing atoms
for idx in idxtoremove:
    merged_final.RemoveAtom(idx)
print("merged molecules here:")
display(merged_final)
display(Chem.MolFromSmiles(Chem.MolToSmiles(merged_final)))
```
giving the below
![image](https://user-images.githubusercontent.com/66372095/138076229-0c68c93b-08d3-4f72-b119-cd5193a61fb3.png)
